### PR TITLE
[autolinking] Patch react podspec for swift integration

### PR DESCRIPTION
**Pull request by Kudo**
# Why

there are tons of mixing swift and objc issues in expo-modules-core. e.g.
- `include of non-modular header inside framework module` when prebuild expo-modules-core
- sometimes swift functions cannot be loaded if there are react types.

the root cause is that clang doesn't generate submodules correctly. although clang can transform `#import <React/RCTBridge.h>` to `@import React.RCTBridge;`, clang cannot do it right as React.RCTBridge submodule doesn't exist. as swift only know modules, this will cause lots of problems when we use react types in swift.

# How

- patch `React-Core.podspec` with custom modulemap for clang to generate submodules correctly. (details in code comment)
- replace original CocoaPods `PodTarget` monkey patch with `Sandbox`. this new approach can patch podspec on the fly.

# Test Plan

- CI passed
- to further check what's the difference for the change, you can enable `BUILD_LIBRARY_FOR_DISTRIBUTION` for expo-modules-core pod project and check the swiftinterface for ExpoModulesCore.
e.g. `/Users/kudo/Library/Developer/Xcode/DerivedData/BareExpo-brnnteedtepltiasryrqyexlvweu/Build/Products/Debug-iphonesimulator/ExpoModulesCore/ExpoModulesCore.swiftmodule/x86_64-apple-ios-simulator.swiftinterface`
```swift
@_inheritsConvenienceInitializers @objc final public class ViewModuleWrapper : ExpoModulesCore.RCTViewManager {
```
before: it's `ExpoModulesCore.RCTViewManager`

```swift
@_inheritsConvenienceInitializers @objc final public class ViewModuleWrapper : React.RCTViewManager {
```
after: it's `React.RCTViewManager`

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [x] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-modules-autolinking/scripts/ios/cocoapods/sandbox.rb [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    private
+
+    _original_store_podspec = instance_method(:store_podspec)
+
+    public
+
+    define_method(:store_podspec) do |name, podspec, _external_source, json|
+
+      # Patch `React-Core.podspec` for clang to generate correct submodules for swift integration
+      if name == 'React-Core'
```
Comment (tsapeta) [inline packages/expo-modules-autolinking/scripts/ios/cocoapods/sandbox.rb:17]: Both names can be versioned in Expo Go, so we should rather check if it contains these names

File: packages/expo-modules-autolinking/scripts/ios/React-Core.modulemap [hunk 2]
```diff
@@ -0,0 +1,6 @@
+module React {
+  umbrella "../../Public/React-Core/React"
```
Comment (tsapeta) [inline packages/expo-modules-autolinking/scripts/ios/React-Core.modulemap:2]: Also here 😞 Shouldn't we version it? I guess maybe we don't have to, but we need to verify that.

File: packages/expo-modules-autolinking/scripts/ios/cocoapods/sandbox.rb (hunk 1 above)
Comment (Kudo) [reply packages/expo-modules-autolinking/scripts/ios/cocoapods/sandbox.rb:17]: good catch! for Expo Go, i used a slight different approach which to generate patched podspec when versioning. the main reason is that we don't have the ABI43_0_0 prefix concept in `expo-modules-autolinking`.

File: packages/expo-modules-autolinking/scripts/ios/React-Core.modulemap (hunk 2 above)
Comment (Kudo) [reply packages/expo-modules-autolinking/scripts/ios/React-Core.modulemap:2]: yep, this is also necessary. thanks!

**Review by tsapeta — APPROVED** [review]

**Comment by Kudo** [comment]
there was an issue to have a different checksum on CI. the module_map is an absolute path and my path is different to CI. so i've add further patch `sandbox.rb` to keep original the checksum. the CI is green now, i'm going to merge this and working on the react delegate pr.

**Pull request merged**